pkg/dns/dns.go

```diff
@@ -0,0 +1,452 @@
+
+	deployment, err := c.client.KubernetesClient().AppsV1().Deployments(metav1.NamespaceSystem).Get("coredns", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	coreConfigMap, err := c.getCorefileConfigMap(deployment)
+	if err != nil {
+		return err
+	}
+
+	c.log.Debug("Patching CoreDNS configmap")
+
+	if err := c.patchCoreDNSConfigMap(coreConfigMap, clusterDomain, maeshNamespace, deployment.Namespace); err != nil {
+		return err
+	}
+
+	if err := c.restartPods(deployment); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) patchCoreDNSConfigMap(coreConfigMap *corev1.ConfigMap, clusterDomain, maeshNamespace, coreNamespace string) error {
+	serverBlock := fmt.Sprintf(
+		`
+%[4]s
+maesh:53 {
+    errors
+    rewrite continue {
+        name regex ([a-zA-Z0-9-_]*)\.([a-zv0-9-_]*)\.maesh %[3]s-{1}-6d61657368-{2}.%[3]s.svc.%[1]s
+        answer name %[3]s-([a-zA-Z0-9-_]*)-6d61657368-([a-zA-Z0-9-_]*)\.%[3]s\.svc\.%[2]s {1}.{2}.maesh
+    }
+    kubernetes %[1]s in-addr.arpa ip6.arpa {
+        pods insecure
+        upstream
+    	fallthrough in-addr.arpa ip6.arpa
+    }
+    forward . /etc/resolv.conf
+    cache 30
+    loop
+    reload
+    loadbalance
+}
+%[5]s
+`,
+		clusterDomain,
+		strings.Replace(clusterDomain, ".", "\\.", -1),
+		maeshNamespace,
+		coreFileHeader,
+		coreFileTrailer,
+	)
+
+	originalBlock := coreConfigMap.Data["Corefile"]
+
+	if strings.Contains(originalBlock, coreFileHeader) {
+		// Corefile already contains the maesh block.
+		return nil
+	}
+
+	newBlock := originalBlock + serverBlock
+	coreConfigMap.Data["Corefile"] = newBlock
+
+	if _, err := c.client.KubernetesClient().CoreV1().ConfigMaps(coreNamespace).Update(coreConfigMap); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getCorefileConfigMap returns the name of a coreDNS config map.
+func (c *Client) getCorefileConfigMap(coreDeployment *appsv1.Deployment) (*corev1.ConfigMap, error) {
+	for _, volume := range coreDeployment.Spec.Template.Spec.Volumes {
+		if volume.ConfigMap == nil {
+			continue
+		}
+
+		cfgMap, err := c.client.KubernetesClient().CoreV1().ConfigMaps(coreDeployment.Namespace).Get(volume.ConfigMap.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		if cfgMap.Data == nil {
+			continue
+		}
+
+		if _, exists := cfgMap.Data["Corefile"]; !exists {
+			continue
+		}
+
+		return cfgMap, nil
+	}
+
+	return nil, errors.New("corefile configmap not found")
+}
+
+// ConfigureKubeDNS patches the KubeDNS configuration for Maesh.
+func (c *Client) ConfigureKubeDNS(maeshNamespace string) error {
+	c.log.Debug("Patching KubeDNS")
+
+	deployment, err := c.client.KubernetesClient().AppsV1().Deployments(metav1.NamespaceSystem).Get("kube-dns", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	var (
+		serviceIP string
+		ebo       = backoff.NewConstantBackOff(10 * time.Second)
+	)
+
+	c.log.Debug("Getting CoreDNS service IP")
+
+	if err = backoff.Retry(safe.OperationWithRecover(func() error {
```

Comment:
Thoughts. If I correctly understand the code we are waiting for the CoreDNS service to be available (service installed by our Helm chart if `kubedns` option is true). 

Maybe we can remove the duplicated CoreDNS config for Maesh and apply a patch here? or create a `configMap` with our config that could be used by the CoreDNS deployment? Does it make sense?

Reply:
Not sure if I understand.

This IP is the service IP of our coreDNS configuration, that we need to have added as a stubDomain in kubeDNS.

We need this IP to have been assigned before we can successfully patch the kubeDNS configmap to forward requests.

Reply:
I'm just saying that we could reuse the logic to patch our CoreDNS deployment which is added as a stubDomain in kubeDNS. This will avoid us to have a duplicate CoreDNS config, one in the code and one in the Helm chart.

Does it make sense?

Reply:
Ohhh, so we deploy ours with an "empty" config...and then patch both our coreDNS and the existing kubeDNS?

hmmmm yep, I like that idea

Reply:
https://github.com/containous/maesh/issues/593